testgrid/cmd/updater/main.go

```diff
@@ -18,14 +18,343 @@ package main
+		// Timed out
+		return state.Row_FAIL
+	default:
+		return state.Row_RUNNING
+	}
+}
+
+func AppendResult(row *state.Row, result state.Row_Result, count int) {
+	latest := int32(result)
+	n := len(row.Results)
+	switch {
+	case n == 0, row.Results[n-2] != latest:
+		row.Results = append(row.Results, latest, int32(count))
+	default:
+		row.Results[n-1] += int32(count)
+	}
+}
+
+func AppendColumn(grid *state.Grid, rows map[string]*state.Row, build BuildResult) {
+	c := state.Column{
+		Build:   build.Id,
+		Started: float64(build.Started * 1000),
+	}
+	grid.Columns = append(grid.Columns, &c)
+
+	missing := map[string]*state.Row{}
+	for name, row := range rows {
+		missing[name] = row
+	}
+
+	for name, result := range build.Results {
+		delete(missing, name)
+		r, ok := rows[name]
+		if !ok {
+			r = &state.Row{
+				Name: name,
+				Id:   name,
+			}
+			rows[name] = r
+			grid.Rows = append(grid.Rows, r)
+			if n := len(grid.Columns); n > 0 {
+				// Add missing entries for later builds
+				AppendResult(r, state.Row_NO_RESULT, n-1)
+			}
+		}
+
+		AppendResult(r, result, 1)
+	}
+
+	for _, row := range missing {
+		AppendResult(row, state.Row_NO_RESULT, 1)
+	}
+}
+
+func ReadBuild(build Build) (*BuildResult, error) {
+	br := BuildResult{
+		Id: path.Base(build.Prefix),
+	}
+	s := build.Bucket.Object(build.Prefix + "started.json")
+	sr, err := s.NewReader(build.Context)
+	if err != nil {
+		return nil, fmt.Errorf("build has not started")
+	}
+	var started Started
+	if err = json.NewDecoder(sr).Decode(&started); err != nil {
+		return nil, fmt.Errorf("could not decode started.json: %v", err)
+	}
+	br.Started = started.Timestamp
+
+	f := build.Bucket.Object(build.Prefix + "finished.json")
+	fr, err := f.NewReader(build.Context)
+	if err == storage.ErrObjectNotExist {
+		return &br, nil
+	}
+
+	var finished Finished
+	if err = json.NewDecoder(fr).Decode(&finished); err != nil {
+		return nil, fmt.Errorf("could not decode finished.json: %v", err)
+	}
+
+	br.Finished = finished.Timestamp
+	br.Passed = finished.Passed
+
+	re := regexp.MustCompile(`.+/junit_.+\.xml$`)
+
+	ai := build.Bucket.Objects(build.Context, &storage.Query{Prefix: build.Prefix + "artifacts/"})
+	var artifacts []string
+	for {
+		a, err := ai.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to list artifacts: %v", err)
+		}
+
+		if !re.MatchString(a.Name) {
+			continue
+		}
+		artifacts = append(artifacts, a.Name)
+	}
+	br.Results = map[string]state.Row_Result{
+		"Overall": br.Overall(),
+	}
+	for _, ap := range artifacts {
+		ar, err := build.Bucket.Object(ap).NewReader(build.Context)
+		if err != nil {
+			return nil, fmt.Errorf("could not read %s: %v", ap, err)
+		}
+		if r := ar.Remain(); r > 50e6 {
+			return nil, fmt.Errorf("too large: %s is %d > 50M", ap, r)
+		}
+		buf, err := ioutil.ReadAll(ar)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read all of %s: %v", ap, err)
+		}
+
+		var suites JunitSuites
```

Comment:
Can you extract this XML parsing (buf -> suite data) and add some unit tests?

Look a gubernator/view_build_test.py for some examples (`test_testsuites` and `test_nested_testsuites`).

Reply:
Done

Please note that this is not expected to create correct grid messages yet (that is coming in later PRs), and a lot of this is gets refactored (note that we throw away the failure message right now, for example)....